components/script/dom/bindings/structuredclone.rs

```diff
@@ -241,112 +277,144 @@ static STRUCTURED_CLONE_CALLBACKS: JSStructuredCloneCallbacks = JSStructuredClon
     canTransfer: Some(can_transfer_callback),
 };
 
-struct StructuredCloneHolder {
-    blob: Option<DomRoot<Blob>>,
+/// A data holder for results from, and inputs to, structured-data read/write operations.
+/// https://html.spec.whatwg.org/multipage/#safe-passing-of-structured-data
+pub enum StructuredDataHolder {
+    Read {
+        /// A deserialized blob, stored temporarily here to keep it rooted.
+        blob: Option<DomRoot<Blob>>,
+        /// A vec of transfer-received DOM ports,
+        /// to be made available to script through a message event.
+        message_ports: Option<Vec<DomRoot<MessagePort>>>,
+        /// A map of port implementations,
+        /// used as part of the "transfer-receiving" steps of ports,
+        /// to produce the DOM ports stored in `message_ports` above.
+        port_impls: Option<HashMap<MessagePortId, MessagePortImpl>>,
```

Comment:
Is there any reason we need the Option wrappers around Vec and HashMap?

Reply:
The only reason I can think of, is that this data structure is used in every operation, regardless of whether ports, blobs, and so on(we will be adding to this list in the future) are actually transferred or serialized, hence the option could be a way to highlight what type of ready/write operation is ongoing. 

Note also that we cannot break up the enum into more specific variants, since we cannot know in advance what operations will be performed in the callbacks. 

On the other hand, we could simply have empty values, versus optional ones.